include/mapbox/vector_tile.hpp

```diff
@@ -207,58 +208,86 @@ std::uint32_t feature::getVersion() const {
 
 template <typename GeometryCollectionType>
 GeometryCollectionType feature::getGeometries(float scale) const {
-    uint8_t cmd = 1;
-    uint32_t length = 0;
-    int32_t x = 0;
-    int32_t y = 0;
+    std::uint8_t cmd = 1;
+    std::uint32_t length = 0;
+    std::int32_t x = 0;
+    std::int32_t y = 0;
 
     GeometryCollectionType paths;
 
     paths.emplace_back();
-    auto* path = &paths.back();
 
     auto start_itr = geometry_iter.begin();
     const auto end_itr = geometry_iter.end();
+    bool first = true;
+    std::uint32_t len_reserve = 0;
+    std::size_t extra_coords = 0;
+    if (type == GeomType::LINESTRING) {
+        extra_coords = 1;
+    } else if (type == GeomType::POLYGON) {
+        extra_coords = 2;
+    }
+    bool is_point = type == GeomType::POINT;
+
     while (start_itr != end_itr) {
         if (length == 0) {
-            uint32_t cmd_length = static_cast<uint32_t>(*start_itr++);
+            std::uint32_t cmd_length = static_cast<std::uint32_t>(*start_itr++);
             cmd = cmd_length & 0x7;
-            length = cmd_length >> 3;
+            length = len_reserve = cmd_length >> 3;
         }
 
         --length;
 
         if (cmd == CommandType::MOVE_TO || cmd == CommandType::LINE_TO) {
-            x += protozero::decode_zigzag32(static_cast<uint32_t>(*start_itr++));
-            y += protozero::decode_zigzag32(static_cast<uint32_t>(*start_itr++));
 
-            if (cmd == CommandType::MOVE_TO && !path->empty()) {
+            if (is_point) {
+                if (first && cmd == CommandType::MOVE_TO) {
+                    // note: this invalidates pointers. So we always
+                    // dynamically get the path with paths.back()
+                    paths.reserve(len_reserve);
+                    first = false;
+                }
+            } else {
+                if (first && cmd == CommandType::LINE_TO) {
+                    paths.back().reserve(len_reserve + extra_coords);
+                    first = false;
+                }
+            }
+
+            if (cmd == CommandType::MOVE_TO && !paths.back().empty()) {
                 paths.emplace_back();
-                path = &paths.back();
+                if (!is_point) first = true;
             }
 
-            float px = std::round(x * float(scale));
-            float py = std::round(y * float(scale));
+            x += protozero::decode_zigzag32(static_cast<std::uint32_t>(*start_itr++));
+            y += protozero::decode_zigzag32(static_cast<std::uint32_t>(*start_itr++));
+            float px = std::round(static_cast<float>(x) * scale);
```

Comment:
Should this be `std::floor` rather then `std::round` as we are checking against what we would static cast and truncate below?